### PR TITLE
Refactor and Document LogFactory class

### PR DESCRIPTION
**Pull request by ie-zero**
* Fix the order of Using statements in LogFactory class. 
* Change the inner classes of LogFactory class to sealed.
* Reduce the visibility of the inner LogFactory.LoggerCacheKey class and convert it to immutable.
* Document the public properties and methods of inner classes (LoggerCacheKey and LoggerCache) in LogFactory.
* Document the fact the some methods in LogFactory class are marked as internal only for Unit Testing purposes.

### CODE REVIEW

**Comment by 304NotModified** [comment]
Looks good! 

What's the benefit of automatic property => getter with (readonly)  backend?

**Comment by ie-zero** [comment]
I my opinion it makes the intention explicit. No other benefit from execution point of view.

**Comment by 304NotModified** [comment]
Well I think I prefer `{get;private set;}` as it is a lot shorter (and less code = less bugs). Note, you also used  `{get;private set;}` while you could used the readonly backend in https://github.com/NLog/NLog/pull/930 ;)

File: src/NLog/LogFactory.cs [hunk 1]
```diff
@@ -1026,35 +1077,28 @@ public Logger Retrieve(LoggerCacheKey cacheKey)
...
-
-                foreach (WeakReference loggerReference in loggerCache.Values)
+                get 
                 {
-                    Logger logger = loggerReference.Target as Logger;
-                    if (logger != null)
-                    {
-                        values.Add(logger);
-                    }
+                    var values =
```
Comment (304NotModified) [inline src/NLog/LogFactory.cs]: Please use LINQ method syntax. `loggerCache.Values.Where(...).Select()` etc.

**Comment by 304NotModified** [comment]
Sorry, If have created an conflict. 

I also have to agree that refactoring is nice, but we should keep it small. It's really difficult to check if it still behaves the same (and the unit tests are not covering 100%)

**Comment by ie-zero** [comment]
This PR will be clean up and re-introduced.

**Comment by 304NotModified** [comment]
:+1:

**Pull request closed without merging**